apps/gdal_translate_lib.cpp

```diff
@@ -3415,15 +3427,55 @@ GDALTranslateOptionsNew(char **papszArgv,
         if (auto aosOutSize =
                 argParser->present<std::vector<std::string>>("-outsize"))
         {
-            if ((*aosOutSize)[0].back() == '%')
-                psOptions->dfOXSizePct = CPLAtofM((*aosOutSize)[0].c_str());
+            const auto &sOutSizeX = (*aosOutSize)[0];
+            const auto &sOutSizeY = (*aosOutSize)[1];
+
+            if (sOutSizeX.back() == '%')
+            {
+                char *end = nullptr;
+                psOptions->dfOXSizePct = CPLStrtodM(sOutSizeX.c_str(), &end);
+                if (*end != '%')
+                {
+                    throw std::invalid_argument(
+                        "Failed to parse first value of -outsize: " +
+                        sOutSizeX);
+                }
+            }
             else
-                psOptions->nOXSizePixel = atoi((*aosOutSize)[0].c_str());
+            {
+                size_t pos;
+                psOptions->nOXSizePixel = std::stoi(sOutSizeX, &pos);
```

Comment:
std::stoi itself can throw an exception, so there are maybe situations where the below informative exception couldn't be reached and a raw one thrown
Same remark for nOYSizePixel

Reply:
How would you feel about the following?

```
            if (sOutSizeX.back() == '%')
            {
                psOptions->dfOXSizePct = parse_or_throw<double>(
                    sOutSizeX, "Failed to parse first value of -outsize");
            }
            else
            {
                psOptions->nOXSizePixel = parse_or_throw<int>(
                    sOutSizeX, "Failed to parse first value of -outsize");
            }

            if (sOutSizeY.back() == '%')
            {
                psOptions->dfOYSizePct = parse_or_throw<double>(
                    sOutSizeY, "Failed to parse second value of -outsize");
            }
            else
            {
                psOptions->nOYSizePixel = parse_or_throw<int>(
                    sOutSizeY, "Failed to parse second value of -outsize");
            }
```

with

```
/**
* Read a single value of a given type from a string_view.
*
* An invalid_value exception will be thrown if the value:
* - cannot be read
* - does not occupy the entire string_view (i.e, if there are trailing non-numeric characters)
* - overflows the desired type
*
* The exception text will be constructed from the provided message base and the unparseable value.
*
* @param s text from which a value should be read
* @param message_base Text used to construct an exception message in case of failure
* @return a value of the desired type
*/
template <typename T>
T parse_or_throw(std::string_view s, std::string_view message_base)
{
    T value;

    if constexpr (std::is_integral_v<T>)
    {
        auto [ptr, ec] = std::from_chars(s.begin(), s.end(), value);

        if (ec == std::errc() && ptr == s.end())
        {
            // no error
            return value;
        }
    }
    else
    {
        // Unfortunately we need to make a copy to to interface with CPLStrtodM
        // because we cannot guarantee that the string_view is null-terminated.
        std::string s_str(s);

        char *end;
        double parsed = CPLStrtodM(s_str.c_str(), &end);
        if (*end == '\0')
        {
            // no error
            if constexpr (std::is_same_v<T, double>)
            {
                return parsed;
            }

            if (parsed >= std::numeric_limits<T>::lowest() &&
                parsed <= std::numeric_limits<T>::max())
            {
                return static_cast<T>(parsed);
            }
        }
    }

    std::ostringstream ss;
    ss << message_base << ": " << s;
    throw std::invalid_argument(ss.str());
}
```

Reply:
LGTM